### ISSUE TITLE
[JOSS] how to set up the mlpack Python bindings

### ISSUE THREAD

**Issue by rasbt** [mlpack/mlpack#1416]
I followed the installation instructions and everything went super smoothly thus far! :)

I can use the binaries from within the bin folder, and also the tests all pass:

```bash
~/mlpack-3.0.1/build$ bin/mlpack_test
Running 1351 test cases...

...
warning: solve(): system appears singular (rcond: 1.31214e-18); attempting approximate solution

warning: solve(): system appears singular (rcond: 1.31214e-18); attempting approximate solution

*** No errors detected
```

I am wondering now what step would be required to enable the Python bindings? Would I just need to add `~/mlpack-3.0.1/src/mlpack/bindings/python/mlpack` to my Python `PATH`? It would be nice if that could be briefly explained in the documentation (maybe in the "building from source" docs).

**Comment by rcurtin** [comment]
Honestly I'd say `make install` is the easiest way to do it, like the build directions suggest.  If you really want to do it manually in the build directory it's hard enough that I wanted to avoid adding it to the documentation... from the build directory:

```
export LD_LIBRARY_PATH=lib/
export PYTHONPATH=src/mlpack/bindings/python/
```

If you `make install` it will put it in `/usr/local/lib/pythonX.Y/site-packages/` which should automatically be in the `PYTHONPATH`.

As I write this though, I do think you are right, it would be useful to add a section to the build documentation mentioning how to work with the Python bindings without installing.

**Comment by rasbt** [comment]
> If you make install it will put it in /usr/local/lib/pythonX.Y/site-packages/ which should automatically be in the PYTHONPATH.

Hm, I used that, but for some reason, it didn't add anything to the site-packages directory (checked both the python 2.7 and 3.5 dirs), which is why I assumed a manual step might be necessary (I used `sudo make install`)

```python
>>> import mlpack
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named mlpack
```

> As I write this though, I do think you are right, it would be useful to add a section to the build documentation mentioning how to work with the Python bindings without installing. I'll do that when I have a chance.

Thanks, that would be nice.

**Comment by rcurtin** [comment]
That's strange. Mind posting the output of `make install` (it will be long...)?  The Python bindings are pretty fiddly and I've found that different systems behave pretty differently.

(The Python bindings are also available via pip for manylinux platforms but where's the fun in that?? :))

**Comment by rcurtin** [comment]
Hmm it seems like the Python bindings aren't actually referenced in that install message.  Does 'make python' work?  It seems like CMake may not have configured the project to build the Python bindings, which could happen if the right Python dependencies aren't available.

**Comment by rasbt** [comment]
> Does 'make python' work?

no, it doesn't seem to work either

```
make: *** No rule to make target 'python'.  Stop.
```

For the whole installation procedure, I used a fresh Ubuntu install (16.04 LTS) and followed the docs step by step, including

```
$ sudo apt-get install libboost-math-dev libboost-program-options-dev
  libboost-test-dev libboost-serialization-dev libarmadillo-dev binutils-dev
  python-pandas python-numpy cython python-setuptools
```

Usually, I use the Miniconda Python distribution, however, for this review, I thought it would be better to follow the installation step by step without any dependencies installed that are not listed. 

Do you think the Cmake step doesn't work with the system's Python, and is there any good way to check that?

**Comment by rcurtin** [comment]
Right, in this case, it looks like CMake didn't find everything right for the Python bindings.  I wonder if the Cython version is too old or something.  Can you run `cmake ../` in the build directory and print the output?  It should show some information about the Python bindings and whether the dependencies could be detected.

**Comment by rcurtin** [comment]
Actually, I bet the Cython version is too old:
https://packages.ubuntu.com/xenial/cython

The Cython version needs to be 0.24 or newer, so you could try on Ubuntu 17.10, but unfortunately 16.04 is pretty old at this point (or nobody has backported a newer version yet).

**Comment by rasbt** [comment]
You are right!

> CMake Warning at CMake/FindPythonModule.cmake:44 (message):
  Unsuitable version of Python module Cython (0.24 or greater required).

Will try a newer Ubuntu version then

**Comment by rasbt** [comment]
Just got one step further ... using Ubuntu 18 seemed to have solved the Cython dependency issue. Sorry, but I know face the next issue. It seems to be a library linking issue when I try to import mlpack in Python:

```python
>>> import mlpack
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/mlpack-3.0.1-py2.7-linux-x86_64.egg/mlpack/__init__.py", line 11, in <module>
    from .test_python_binding import test_python_binding
ImportError: libmlpack.so.3: cannot open shared object file: No such file or directory
```

I just `grep`ed through the `cmake`, `make`, and `make install` outputs and couldn't find anything suspicious there; I pasted the relevant lines below. Have you encountered this or a similar issue in the past and have any tips/recommendation for resolving this?

```
sudo cmake ../
-- cotire 1.7.10 loaded.
-- Armadillo libraries: /usr/lib/libarmadillo.so
-- Boost version: 1.65.1
-- Found the following Boost libraries:
--   program_options
--   unit_test_framework
--   serialization
-- Regenerating arma_config.hpp.
-- Found Python: /usr/bin/python
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
CMake Warning at CMakeLists.txt:487 (message):
  txt2man not found; man pages will not be generated.


-- Configuring done
-- Generating done
-- Build files have been written to: /home/ubuntu/mlpack-3.0.1/build
ubuntu@ip-172-31-14-44:~/mlpack-3.0.1/build$ sudo make
[  1%] Updating arma_config.hpp (if necessary)
-- Regenerating arma_config.hpp.
[  1%] Built target mlpack_arma_config
Moving header files to include/mlpack/
[  1%] Built target mlpack_headers
[  2%] Generating mlpack.pc (pkg-config) file.
[  2%] Built target pkgconfig
[  2%] Building CXX precompiled header src/mlpack/cotire/mlpack_CXX_prefix.hxx.gch


...


[ 16%] Linking CXX shared library ../../lib/libmlpack.so
[ 16%] Built target mlpack
running build
running build_py
copying /home/ubuntu/mlpack-3.0.1/build/src/mlpack/bindings/python/mlpack/__init__.py -> build/lib.linux-x86_64-2.7/mlpack
copying /home/ubuntu/mlpack-3.0.1/build/src/mlpack/bindings/python/mlpack/matrix_utils.py -> build/lib.linux-x86_64-2.7/mlpack
[ 16%] Built target python_copy
Scanning dependencies of target generate_pyx_sparse_coding
[ 16%] Building CXX object src/mlpack/methods/sparse_coding/CMakeFiles/

...
```

```
~/mlpack-3.0.1/build$ sudo make
...

[ 15%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/sparse_autoencoder/sparse_autoencoder_function.cpp.o
[ 15%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/sparse_autoencoder/maximal_inputs.cpp.o
[ 15%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/sparse_coding/sparse_coding.cpp.o
[ 16%] Linking CXX shared library ../../lib/libmlpack.so
[ 16%] Built target mlpack
running build
running build_py
copying /home/ubuntu/mlpack-3.0.1/build/src/mlpack/bindings/python/mlpack/__init__.py -> build/lib.linux-x86_64-2.7/mlpack
copying /home/ubuntu/mlpack-3.0.1/build/src/mlpack/bindings/python/mlpack/matrix_utils.py -> build/lib.linux-x86_64-2.7/mlpack
[ 16%] Built target python_copy
...
```


```
~/mlpack-3.0.1/build$ sudo make install
[  1%] Updating arma_config.hpp (if necessary)
[  1%] Built target mlpack_arma_config

...

-- Installing: /usr/local/include/mlpack/methods/softmax_regression/softmax_regression.hpp
-- Installing: /usr/local/include/mlpack/methods/softmax_regression/softmax_regression_impl.hpp
-- Installing: /usr/local/include/mlpack/prereqs.hpp
-- Installing: /usr/local/include/mlpack/core.hpp
-- Installing: /usr/local/lib/libmlpack.so.3.0
-- Up-to-date: /usr/local/lib/libmlpack.so.3
-- Up-to-date: /usr/local/lib/libmlpack.so
running install
running bdist_egg
running egg_info
...
```


PS: `apt-get install pkg-config` could be added to the dependency list, i.e., in the lines

```
# apt-get install libboost-math-dev libboost-program-options-dev
  libboost-test-dev libboost-serialization-dev libarmadillo-dev binutils-dev
  python-pandas python-numpy cython python-setuptools
```
since it doesn't come with Ubuntu by default.

**Comment by rcurtin** [comment]
If it's installed to `/usr/local/`, on some systems (including Ubuntu) this is not in the default linker search path.  So you'll need to `export LD_LIBRARY_PATH=/usr/local/lib/` since that's where libmlpack.so.3 is.

pkg-config isn't a requirement for building, though, it's optional---so if it's not there, no problem.  (You just can't use pkg-config later to link against mlpack, for instance.)

**Comment by rcurtin** [comment]
(more details in the README at the end of section 4: https://github.com/mlpack/mlpack#4-building-mlpack-from-source)

**Comment by rasbt** [comment]
Good point, thanks! I usually use CentOS, and I think it's the default there neither (prob. why I have that line looming around in my .bashrc). 

Anyways it seems to work now :). Could you add the last few lines of Section 4 to the online docs as well? I think that's very helpful info for users since Ubuntu is one of the recommended (or "example") distros. Don't want to be too picky, but then I can check off the "Installation: Does installation proceed as outlined in the documentation?" review item in good consciousness :)

**Comment by rcurtin** [comment]
That's a fair change and I would agree with it.  I'll incorporate it with the other changes I said I would do earlier.

**Comment by rcurtin** [comment]
These should be fixed when #1421 is merged.  Thanks again for pointing the issues out, I really appreciate your dive in.  Many times it can be hard to get good feedback from an outsider's perspective.

**Comment by rcurtin** [comment]
Fixed via #1421.
